### ISSUE TITLE
some CatchCNCLM45 restart variables are not processed correctly in ScaleCatchCN.F90 and CatchmentCNRst.F90

### ISSUE THREAD

**Issue by gmao-jkolassa** [GEOS-ESM/GEOSgcm_GridComp#651]
In ScaleCatchCN.F90 and CatchmentCNRst.F90 some CatchCNCLM45-specific restart variables are written to the output restart file, but are not read from the input restart file. The effect of this is that these variables are set to 0 everywhere in the output restart file. 

The variables affected are:

- RHM
- WINDM
- RAINFM
- SNOWFM
- RUNSURFM
- AR1M
- T2M10D
- TPREC10D
- TPREC60D

This effect only manifests when the  resolution of the input restart file differs from that of the output restart file. 

@weiyuan-jiang @gmao-rreichle

**Comment by gmao-jkolassa** [comment]
Just a follow-up comment to say that the setting to 0 is less of an issue for the variables that end in "M" (they are CN time step averages), but should definitely not be done for the variables ending in "D". 

It also makes me wonder why the variables that are set to 0 are included as restart variables if they are just generally set to 0. 

And finally, whatever is implemented going forward, I think it would be good to ensure consistency. Currently, the variables are only set to 0 when the resolution of the input and output restart variable differs, but the values from the input file are inherited when the files have the same resolution. (Maybe this point has already been addressed in the revision of the mk_restarts utilities?)

**Comment by gmao-rreichle** [comment]
> Just a follow-up comment to say that the setting to 0 is less of an issue for the variables that end in "M" (they are CN time step averages), but should definitely not be done for the variables ending in "D".

@gmao-jkolassa, I don't really know anything about CatchCN restarts, but from your description it sounds like there's a difference between a remapped restart that will require spin-up and a restart that can be used to extend a simulation.  After remapping the restart, there's some spinup needed anyway, and this means that not dealing with the running mean values of the forcing is ok in this case.  
> It also makes me wonder why the variables that are set to 0 are included as restart variables if they are just generally set to 0.

This sounds like a bug to me.  If we want to extend a simulation, all restart variables should have exactly the value that they had at the end of the preceding time step.  As implemented right now, I suspect the code might not pass a simple regression test whereby a say, 1-month long simulation is done in one go vs. another that stops and starts once a day. 

> And finally, whatever is implemented going forward, I think it would be good to ensure consistency. Currently, the variables are only set to 0 when the resolution of the input and output restart variable differs, but the values from the input file are inherited when the files have the same resolution. (Maybe this point has already been addressed in the revision of the mk_restarts utilities?)

I doubt that this CatchCN issue was fixed in the recent revision of mk_restarts.  This is all news to me. 

cc: @rdkoster

**Comment by gmao-rreichle** [comment]
@gmao-jkolassa: I thought you mentioned that some variables are put into the restart file (as non-zero) but are then not read from the restart file and thus effectively reinitialized to zero.  Maybe I misunderstood your comments.  But if what I thought is indeed the case, then CatchCNCLM45 would **not** pass the above-mentioned regression test.  That is, CatchCNCLM45 would give you a different result depending on when and how often it is stopped and restarted.  Can you please clarify?

**Comment by gmao-jkolassa** [comment]
> @gmao-jkolassa: I thought you mentioned that some variables are put into the restart file (as non-zero) but are then not read from the restart file and thus effectively reinitialized to zero. Maybe I misunderstood your comments. But if what I thought is indeed the case, then CatchCNCLM45 would **not** pass the above-mentioned regression test. That is, CatchCNCLM45 would give you a different result depending on when and how often it is stopped and restarted. Can you please clarify?

This issue only occurs when the restart file used to set up an experiment has a different resolution from the experiment resolution (so if the input and output restart file for ScaleCatchCN.F90 or CatchmentCNRst.F90 have different resolutions). So for example:

Scenario 1: You set up an M36 experiment with an M36 restart file. All fields in the restart file generated by ldas_setup are populated with the values from the input restart file.

Scenario 2: You set up and M36 experiment with an M09 restart files. The fields for the above variables are set to 0 everywhere in the restart file generated by ldas_setup.

Personally, I think it would be better to populate the restart file in scenario 2 with reasonable values rather than 0, but it is maybe less of an issue than I initially thought, because in scenario 2 we would have to run a several months spin-up anyway. 

Does this answer your question?

**Comment by gmao-rreichle** [comment]
@gmao-jkolassa, thanks for the clarification.  I'm glad that what looked like a bug isn't really a true bug.  But I agree that it would be better to remap all of the restart variables listed above and make sure to read them from the remapped restart.  Having these variables processed correctly in the restart should shorten the spin-up that will be needed.  I wonder what the CatchCN code does when the running mean of T2m is near absolute zero everywhere...
Please work with @weiyuan-jiang to create a branch and PR.  Thanks.
cc: @rdkoster

**Comment by weiyuan-jiang** [comment]
There are more variables are set to zero. Can you confirm that are what you need?  Are those variable set to zero intentionally? @gmao-jkolassa 

https://github.com/GEOS-ESM/GEOSgcm_GridComp/blob/775925a7b162379092d7f3e353dd021bf5452277/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/mk_restarts/CatchmentCNRst.F90#L51-L70

**Comment by gmao-jkolassa** [comment]
@weiyuan-jiang I think for consistency it would be best to include all the variables that you listed above. Thanks!

**Comment by gmao-rreichle** [comment]
@gmao-jkolassa : 

1) Are all of the above variables in the "internal specs" **and** actually used in CatchCN?  If they're not used, now would be a good time to clean up.  Also, is there a difference between CatchCNCLM40 and CatchCNCLM45 for these variables?

2) Just to be safe, I think we need to conduct a regression test on CatchCNCLM40 and CatchCNCLM45 as follows:
```
CTRL: Run CatchCNCLM4x for a 2-month period in one segment (not restarting).
EXP:  Run CatchCNCLM4x for the 1st month, then stop and restart to run the 2nd month.
```
At the end of the 2-month experiment period, the restart files from CTRL and EXP must be binary identical.

**Comment by gmao-jkolassa** [comment]
@gmao-rreichle :

`Are all of the above variables in the "internal specs" and actually used in CatchCN? If they're not used, now would be a good time to clean up. Also, is there a difference between CatchCNCLM40 and CatchCNCLM45 for these variables?`

All of these variables are in the internal specs and are used in the code (although some of them may become obsolete in the new CatchCNCLM51). Some of the variables are unique to CatchCNCLM40 while others are unique to CatchCNCLM45. The exact split is as follows:

**Unique to CatchCNCLM40:**
SFMCM

**Unique to CatchCNCLM45:**
AR1M, RAINFM, RHM, RUNSRFM, SNOWFM, WINDM, TPREC10D, TPREC60D, T2M10D

The rest of the variables are used in both versions of CatchCN. 

Because of how the restart utility module is currently constructed all of the variables have to be allocated and added to the catch-type, even if they are only used in one of the model versions. The actual reading and writing of the variables is however nested in if-blocks that distinguish between the two model versions. 

`Just to be safe, I think we need to conduct a regression test on CatchCNCLM40 and CatchCNCLM45 as follows:...`

Ok, I'll go ahead and set up a set of experiments and will report back once they are complete.

**Comment by weiyuan-jiang** [comment]
> 

For this experiment, no remapping is involved. So it should be fine.  Also I think it is not just read and write those variables.  The regrid_carbon subroutines also needs to change.  I may need help on how to re-tiles those variables.

**Comment by gmao-rreichle** [comment]
> For this experiment, no remapping is involved. So it should be fine. 

That's the hope.  The test is to make sure it's indeed the case.  Given the various and lengthy lists of variables, I wouldn't too surprised if something got lost in translation.  I remember that we did such a regression test for Catch a long while ago.  I don't remember hearing of such a test for CatchCN, so it's definitely worth checking.

**Comment by gmao-jkolassa** [comment]
`The regrid_carbon subroutines also needs to change. I may need help on how to re-tiles those variables.`

I think the regrid_carbon subroutine is specifically for the CNCOL and CNPFT arrays, which contain the carbon and nitrogen state variables. The variables listed above are not part of those arrays, so I don't think regrid_carbon needs to be changed. For the most part, the variables above are atmospheric forcing fields or surface states, so it should be fine to process them in the same way as the other non-carbon variables. Exceptions to this might be AR1M, PSNSUNM and PSNSHAM. I will think about how to treat these and get back to you.

**Comment by gmao-jkolassa** [comment]
Following @gmao-rreichle's suggestion I ran the following experiments for CatchmentCNCLM40 and CatchmentCNCLM45.

**CONTROL:** run the model for a two-month period continuously.
**EXP:**            run the model for one month, stop and restart to run the second month

Then I compared the restart files written out after the two months for both CatchCN versions. They were identical for CatchmentCNCLM40, but showed difference in 40 variables an at multiple locations for each variable in the case of CatchmentCNCLM45.
